napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,103 @@
+import logging
+import weakref
+from concurrent.futures import Executor, Future, ThreadPoolExecutor
+from threading import Lock
+from typing import Dict, Iterable, Tuple
+
+from napari.components import Dims
+from napari.layers import Layer
+from napari.utils.events.event import EmitterGroup, Event
+
+LOGGER = logging.getLogger("napari.components._layer_slicer")
+
+
+class _LayerSlicer:
+    def __init__(self):
+        self.events = EmitterGroup(source=self, ready=Event)
+        self._executor: Executor = ThreadPoolExecutor(max_workers=1)
+        self._layers_to_task: Dict[Tuple[Layer], Future] = {}
+        self.lock = Lock()
```

Comment:
This should probably be an `RLock`, which I think should generally be preferred to `Lock`. That should prevent some of the tests from hanging though some may still fail.